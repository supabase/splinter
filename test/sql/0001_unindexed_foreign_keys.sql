begin;

    savepoint a;

    set local search_path = '';

    -- Simple Case
    -- No index on bbb.aaa_id produces an error
    create table public.aaa(
        id int primary key
    );

    create table public.bbb(
        id int primary key,
        aaa_id int references public.aaa(id) -- no index
    );

    select * from lint."0001_unindexed_foreign_keys";


    -- When a covering index is created, the error goes away
    create index on public.bbb(aaa_id);
    select * from lint."0001_unindexed_foreign_keys";


    rollback to savepoint a;


    -- Multi-column Case
    -- No index on bbb(foo, bar)
    create table public.aaa(
        foo int,
        bar int,
        primary key (foo, bar)
    );

    create table public.bbb(
        id int primary key,
        foo int,
        bar int,
        foreign key (foo, bar) references public.aaa(foo, bar)
    );

    select * from lint."0001_unindexed_foreign_keys";

    -- Confirm that an index on the correct columns but in the wrong order
    -- does NOT resolve the issue

    create index on public.bbb(bar, foo);
    select * from lint."0001_unindexed_foreign_keys";

    -- When we create a multi-column index in the correct order the issue is resolved
    create index on public.bbb(foo, bar);
    select * from lint."0001_unindexed_foreign_keys";


rollback;
