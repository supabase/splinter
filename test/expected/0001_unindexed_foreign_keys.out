begin;
    savepoint a;
    set local search_path = '';
    -- Simple Case
    -- No index on bbb.aaa_id produces an error
    create table public.aaa(
        id int primary key
    );
    create table public.bbb(
        id int primary key,
        aaa_id int references public.aaa(id) -- no index
    );
    select * from lint."0001_unindexed_foreign_keys";
          name          |         title          | level |  facing  |  categories   |                                             description                                             |                                                               detail                                                                |                                        remediation                                         |                                                 metadata                                                  |                     cache_key                     
------------------------+------------------------+-------+----------+---------------+-----------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------+---------------------------------------------------
 unindexed_foreign_keys | Unindexed foreign keys | INFO  | EXTERNAL | {PERFORMANCE} | Identifies foreign key constraints without a covering index, which can impact database performance. | Table \`public.bbb\` has a foreign key \`bbb_aaa_id_fkey\` without a covering index. This can lead to suboptimal query performance. | https://supabase.com/docs/guides/database/database-linter?lint=0001_unindexed_foreign_keys | {"name": "bbb", "type": "table", "schema": "public", "fkey_name": "bbb_aaa_id_fkey", "fkey_columns": [2]} | unindexed_foreign_keys_public_bbb_bbb_aaa_id_fkey
(1 row)

    -- When a covering index is created, the error goes away
    create index on public.bbb(aaa_id);
    select * from lint."0001_unindexed_foreign_keys";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

    rollback to savepoint a;
    -- Multi-column Case
    -- No index on bbb(foo, bar)
    create table public.aaa(
        foo int,
        bar int,
        primary key (foo, bar)
    );
    create table public.bbb(
        id int primary key,
        foo int,
        bar int,
        foreign key (foo, bar) references public.aaa(foo, bar)
    );
    select * from lint."0001_unindexed_foreign_keys";
          name          |         title          | level |  facing  |  categories   |                                             description                                             |                                                                detail                                                                |                                        remediation                                         |                                                   metadata                                                    |                     cache_key                      
------------------------+------------------------+-------+----------+---------------+-----------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+----------------------------------------------------
 unindexed_foreign_keys | Unindexed foreign keys | INFO  | EXTERNAL | {PERFORMANCE} | Identifies foreign key constraints without a covering index, which can impact database performance. | Table \`public.bbb\` has a foreign key \`bbb_foo_bar_fkey\` without a covering index. This can lead to suboptimal query performance. | https://supabase.com/docs/guides/database/database-linter?lint=0001_unindexed_foreign_keys | {"name": "bbb", "type": "table", "schema": "public", "fkey_name": "bbb_foo_bar_fkey", "fkey_columns": [2, 3]} | unindexed_foreign_keys_public_bbb_bbb_foo_bar_fkey
(1 row)

    -- Confirm that an index on the correct columns but in the wrong order
    -- does NOT resolve the issue
    create index on public.bbb(bar, foo);
    select * from lint."0001_unindexed_foreign_keys";
          name          |         title          | level |  facing  |  categories   |                                             description                                             |                                                                detail                                                                |                                        remediation                                         |                                                   metadata                                                    |                     cache_key                      
------------------------+------------------------+-------+----------+---------------+-----------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+----------------------------------------------------
 unindexed_foreign_keys | Unindexed foreign keys | INFO  | EXTERNAL | {PERFORMANCE} | Identifies foreign key constraints without a covering index, which can impact database performance. | Table \`public.bbb\` has a foreign key \`bbb_foo_bar_fkey\` without a covering index. This can lead to suboptimal query performance. | https://supabase.com/docs/guides/database/database-linter?lint=0001_unindexed_foreign_keys | {"name": "bbb", "type": "table", "schema": "public", "fkey_name": "bbb_foo_bar_fkey", "fkey_columns": [2, 3]} | unindexed_foreign_keys_public_bbb_bbb_foo_bar_fkey
(1 row)

    -- When we create a multi-column index in the correct order the issue is resolved
    create index on public.bbb(foo, bar);
    select * from lint."0001_unindexed_foreign_keys";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

    rollback to savepoint a;
    -- Issue 90: If a fkey is a subset of an existing index (in the same order) then
    -- it is indexed and should not raise an error
    create table public.tenant(
        id uuid primary key,
        name text not null
    );
    create table public.account(
        tenant_id uuid not null references public.tenant(id),
        user_id uuid not null,
        primary key (tenant_id, user_id)
    );
    select * from lint."0001_unindexed_foreign_keys";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

rollback;
